CMakeLists.txt

```diff
@@ -160,27 +160,73 @@ add_custom_command(
+    set(INCLUDE_JDK17PLUS_DIR FALSE)
+else()
+    set(INCLUDE_JDK17PLUS_DIR TRUE)
+endif()
+
+# macro to filter source lists based on conditions and directory existence checks
+# FILTER_LIST: the list variable to filter
+# PATH_PATTERN: regex pattern to match paths for filtering (".*/jdk17plus/.*")
+# CONDITION: Boolean condition for whether or not the path PATH_PATTERN is excluded
+# CHECK_DIR: Directory path to check for existence - can remove after jdk17plus directory is merged to ACCP 
+macro(filter_sources_by_condition FILTER_LIST PATH_PATTERN CONDITION CHECK_DIR)
+    if(NOT ${CONDITION})
+        list(FILTER ${FILTER_LIST} EXCLUDE REGEX "${PATH_PATTERN}")
+    elseif(CHECK_DIR AND NOT EXISTS "${CHECK_DIR}")
+        list(FILTER ${FILTER_LIST} EXCLUDE REGEX "${PATH_PATTERN}")
+    endif()
+endmacro()
+
 # NOTE: CMake introduced the CONFIGURE_DEPENDS feature in 3.12, so condition
 #       its use on CMake version. If the feature is not available, you may need
 #       to `touch CMakeLists.txt` after adding/removing files for them to be
 #       detected by CMake.
 if (${CMAKE_VERSION} VERSION_LESS "3.12.0")
     file(GLOB_RECURSE ACCP_SRC "src/com/amazon/corretto/crypto/provider/*.java")
     file(GLOB_RECURSE ACCP_UTILS_SRC "src/com/amazon/corretto/crypto/utils/*.java")
+    filter_sources_by_condition(ACCP_SRC ".*/jdk17plus/.*" ${INCLUDE_JDK17PLUS_DIR} "${CMAKE_CURRENT_SOURCE_DIR}/src/com/amazon/corretto/crypto/provider/jdk17plus")
 else()
     file(GLOB_RECURSE ACCP_SRC CONFIGURE_DEPENDS "src/com/amazon/corretto/crypto/provider/*.java")
     file(GLOB_RECURSE ACCP_UTILS_SRC CONFIGURE_DEPENDS "src/com/amazon/corretto/crypto/utils/*.java")
+    filter_sources_by_condition(ACCP_SRC ".*/jdk17plus/.*" ${INCLUDE_JDK17PLUS_DIR} "${CMAKE_CURRENT_SOURCE_DIR}/src/com/amazon/corretto/crypto/provider/jdk17plus")
 endif()
 set(ACCP_SRC ${ACCP_SRC} ${ACCP_UTILS_SRC} ${GENERATED_JAVA_SRC})
 
 set(BASE_JAVA_COMPILE_FLAGS ${CMAKE_JAVA_COMPILE_FLAGS} -h "${JNI_HEADER_DIR}" -Werror)
 
-# Java targets defined here are compiled for Java supporting modules
-set(CMAKE_JAVA_COMPILE_FLAGS ${BASE_JAVA_COMPILE_FLAGS} -Xlint --release 9)
+# This macro sets the Java compile flags for ACCP based on JDK version and compilation type
+# On JDK 8: uses --release 9 for module compilation as modules are supported in JDK 9+, and uses --release 8 for code
+# JDK 17 with KEM support: when targeting JDK 17 on newer build-time JDKs (21+), uses -source/-target instead of --release.
+# This is because using --release 17 would exclude backported APIs and cause runtime errors when accessing KEM functionality or other JDK 17 APIs.
+# On default, it uses the specified target JDK version.
+# Parameters:
+#   FLAGS_VAR - CMake variable name to store the resulting compiler flags
+#   IS_MODULE - TRUE for module compilation, FALSE for regular code
+macro(set_java_compile_flags FLAGS_VAR IS_MODULE)
```

Comment:
I don't think you need to take in `FLAGS_VAR` if it is always set to `CMAKE_JAVA_COMPILE_FLAGS`. This macro signature can just be `macro(set_java_compile_flags IS_MODULE)`